multicast/hear.py

```diff
@@ -332,30 +297,30 @@ class HearUDPHandler(socketserver.BaseRequestHandler):
 	"""
 
 	def handle(self):
-		data = self.request[0].strip()
-		socket = self.request[1]
+		(data, sock) = self.request
+		_sim_data_str = data.strip().replace('\r', '').replace('%', '%%')
```

Comment:
🤔 1. Please assume `self.socket_type == socket.SOCK_DGRAM` please read up on the [official docs on handle function](https://docs.python.org/3.9/library/socketserver.html#socketserver.BaseRequestHandler.handle)
Now, why do you assume there are any bytes in play?

2. That aside will your suggestion work if data is already a string? (ie can we force utf-8 decoding?) Please read the whole file `multicast/hear.py` for more context.

Reply:
😕 1. Um?!? in the linked documentation the example literally uses `data = self.request[0].strip()` which is strong indication that `data` and `self.request[0]` are both strings: otherwise we could expect a `TypeError` or simmilar issue. Further please re-read "https://docs.python.org/3.9/library/socketserver.html#socketserver.BaseRequestHandler.handle" and focus on the handle function. You should note the part where it specifies "for datagram services, self.request is a pair of string and socket". Given that I see a flaw in your reasoning:
> Why I assumed data is of type bytes: Given that `self.socket_type == socket.SOCK_DGRAM`, the `handle` method in a `socketserver.BaseRequestHandler` receives `self.request` as a tuple `(data, socket)`, where `data` is typically a bytes object.
* your conclusion that '`data` is typically a `bytes` object' is a falsehood, whenever  `self.request` is not a `socket` object it is a `tuple` and when it is a `tuple` it is a string and socket pair.

2. I like this approach, but data will never be (except in the erroneous case) be a byte object; is there a way to force utf-8 on a string, perhaps by re-interpreting the string instead?